packages/hardhat-utils/src/fs.ts

```diff
@@ -0,0 +1,546 @@
+ *
+ * @param absolutePath The path to the file or directory to remove.
+ * @throws FileSystemAccessError for any error, except for non-existent path errors.
+ */
+export async function remove(absolutePath: string): Promise<void> {
+  try {
+    await fsPromises.rm(absolutePath, {
+      recursive: true,
+      force: true,
+      maxRetries: 3,
+    });
+  } catch (e) {
+    ensureError<NodeJS.ErrnoException>(e);
+    throw new FileSystemAccessError(e.message, e);
+  }
+}
+
+/**
+ * Changes the permissions of a file or directory.
+ *
+ * @param absolutePath The path to the file or directory.
+ * @param mode The permissions to set. It can be a string or a number representing the octal mode.
+ * @throws FileNotFoundError if the path doesn't exist.
+ * @throws FileSystemAccessError for any other error.
+ */
+export async function chmod(
+  absolutePath: string,
+  mode: string | number,
+): Promise<void> {
+  try {
+    await fsPromises.chmod(absolutePath, mode);
+  } catch (e) {
+    ensureError<NodeJS.ErrnoException>(e);
+    if (e.code === "ENOENT") {
+      throw new FileNotFoundError(absolutePath, e);
+    }
+
+    throw new FileSystemAccessError(e.message, e);
+  }
+}
+
+/**
+ * Creates a file with an empty content. If the file already exists, it will be overwritten.
+ * If part of the path doesn't exist, it will be created.
+ *
+ * @param absolutePath The path to the file to create.
+ * @throws FileSystemAccessError for any other error.
+ */
+export async function createFile(absolutePath: string): Promise<void> {
+  await writeUtf8File(absolutePath, "");
+}
+
+/**
+ * Empties a directory by recursively removing all its content. If the
+ * directory doesn't exist, it will be created. The directory itself is
+ * not removed.
+ *
+ * @param absolutePath The path to the directory to empty.
+ * @throws NotADirectoryError if the path is not a directory.
+ * @throws FileSystemAccessError for any other error.
+ */
+export async function emptyDir(absolutePath: string): Promise<void> {
+  let isDir;
+  let mode;
+  try {
+    const stats = await fsPromises.stat(absolutePath);
+    isDir = stats.isDirectory();
+    mode = stats.mode;
+  } catch (e) {
+    ensureError<NodeJS.ErrnoException>(e);
+    if (e.code === "ENOENT") {
+      await mkdir(absolutePath);
+      return;
+    }
+
+    throw new FileSystemAccessError(e.message, e);
+  }
+
+  if (!isDir) {
+    throw new NotADirectoryError(absolutePath, new Error());
+  }
+
+  await remove(absolutePath);
+  await mkdir(absolutePath);
+  // eslint-disable-next-line no-bitwise
+  await chmod(absolutePath, mode & 0o777);
+}
+
+/**
+ * Looks for a file in the current directory and its parents.
+ *
+ * @param fileName The name of the file to look for.
+ * @param from The directory to start the search from. Defaults to the current working directory.
+ * @returns The absolute path to the file, or `undefined` if it wasn't found.
+ */
+export async function findUp(
+  fileName: string,
```

Comment:
I realized this will not fail if `fileName` is a path, but it may not behave as expected. I would add a check at the start and throw if a path is used.